lib/services/catalog/controllers/catalog-search-controller.js

```diff
@@ -12,20 +13,64 @@ class CatalogSearchController extends Controller {
 	}
 
 	get(req, res, next) {
-		const query = req.query.q;
+		const query = _.get(req, 'query.q', '');
 		const channel = (req.identity.channel || {}).id;
-		const type = req.query.type;
 
 		if (!channel) {
 			return next(Boom.badRequest('JSON Web Token must include a channel ID for search.'));
 		}
 
-		const args = {query, channel, type};
+		if (query.length <= 3) {
+			return next(Boom.badRequest('`q` query paramater must be at least 3 characters'));
+		}
+
+		let types;
+		if ((req.query || {}).types) {
+			// At this time, only allow sorting by one paramater.
+			types = req.query.types.split(',');
+		} else {
+			types = ['video', 'collection'];
+		}
+
+		let sortBy;
+		if ((req.query || {}).sort) {
+			// At this time, only allow sorting by one paramater.
+			sortBy = req.query.sort.split(',')[0];
+		} else {
+			sortBy = null;
+		}
+
+		const page = {
+			offset: parseInt(_.get(req, 'query.page.offset'), 10) || 0,
+			limit: parseInt(_.get(req, 'query.page.limit'), 10) || 10
+		};
+
+		const args = {query, channel, types};
 
 		return this.bus.query({role: 'catalog', cmd: 'search'}, args)
-			.then(objects => {
+			.then(data => {
+				// Sorting
+				if (sortBy && data.length > 0) {
+					// At this time, only allow sorting by one paramater.
+					const descending = sortBy.charAt(0) === '-';
+					const osSortBy = sortBy.replace(/^-/, '');
+
+					data = data.sort(objectSort(osSortBy));
+					if (descending) {
+						data = data.reverse();
+					}
+				}
+
+				// Paging
+				if (page && data.length > 0) {
+					const start = page.offset * page.limit;
```

Comment:
Ok, so I'm wrong above. Since offset defaults to `0`, then `const start = page.offset;`. That means `const end = start + page.limit - 1;`, since slice's end parameter is also a 0-based index.